### PR TITLE
feat(monitor): startup reconciliation for derived events via cursor replay (fixes #1667)

### LINKED ISSUE #1667: feat(monitor): startup reconciliation for work items stuck in qa after crash or downtime
## Problem

The `DerivedEventPublisher` (added in #1658) has no crash recovery. If the daemon crashes between `eventLog.append(pr.merged)` completing and the `prMergedToDone.derive` transaction committing, the work item stays in `qa` permanently — the event log has `pr.merged` but no corresponding `phase.changed`, and the poller won't re-fire since it already marked `prState: merged`.

The same gap exists for daemon downtime: if the daemon was stopped when a PR merged, the next poll will observe `prState: merged` and fire `pr.merged`, which will work correctly — but only if the poller state resets on restart. If the poller uses a persistent cursor (stores last-observed state in SQLite), the merged PR is silently skipped.

## Proposed Fix

On daemon startup, after `DerivedEventPublisher` is initialized, run a reconciliation pass:

```sql
SELECT wi.*
FROM work_items wi
WHERE wi.phase = 'qa'
  AND wi.pr_number IS NOT NULL
  AND EXISTS (
    SELECT 1 FROM monitor_events
    WHERE event = 'pr.merged' AND pr_number = wi.pr_number
  )
```

For each matching row, synthesize and publish a `pr.merged` event (or directly call `prMergedToDone.derive`) so the transition fires. This handles both crash recovery and downtime.

## Bonus: generalizes with a derived cursor

A `derived_cursor` table tracking the last `seq` processed by the publisher would let startup replay any missed events from the event log since the cursor, not just the `qa + pr_merged` join. This also enables future rules to benefit from the same replay mechanism automatically.

## Source

Discovered during adversarial review of #1658.

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary
- Adds a `derived_cursor` table to `DerivedEventPublisher` that tracks the last event `seq` processed, persisted in SQLite
- On daemon startup, `reconcile()` replays any events from the event log since the cursor, ensuring work items stuck in `qa` after a `pr.merged` event (due to crash or downtime) get transitioned to `done`
- Cursor advances atomically inside the same transaction as rule application, so crash between event-log append and derived-rule commit is fully recoverable
- Generalizes beyond `pr.merged → done`: any future derived rules automatically benefit from the same replay mechanism

## Test plan
- [x] Cursor advances as events are processed
- [x] Reconciliation replays missed `pr.merged` and transitions QA work item to done
- [x] No-op when cursor is caught up (idempotent)
- [x] Multiple missed events are all replayed
- [x] Reconcile without eventLog returns 0
- [x] Cursor persists across publisher instances
- [x] All existing derived-events tests still pass (27 total)
- [x] Full test suite passes (5803 tests, 0 failures)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1712 — feat(monitor): startup reconciliation for derived events via cursor replay (fixes #1667)
**Linked issue:** #1667

### Implementation Evidence

- `packages/daemon/src/derived-events.ts`: `DerivedEventPublisher` now creates and manages a `derived_cursor` SQLite table (`migrateCursor`, `loadCursor`, `saveCursor` methods). Cursor advances atomically inside the same transaction as rule application in `handleEvent`, preventing any crash window. `reconcile()` calls `eventLog.getSince(this.cursor)` and re-runs `handleEvent` for each missed event, returning the replay count (0 if no eventLog).
- `packages/daemon/src/index.ts` (lines 942–950): `eventLog` is now passed into `DerivedEventPublisher`; `reconcile()` is called immediately after construction with a log message when events were replayed.
- The approach generalizes: any future `DerivedRule` automatically benefits from replay without additional code.

### Test Coverage

- `packages/daemon/src/derived-events.spec.ts` — 6 new cursor/reconciliation tests:
  - **cursor advances as events are processed** — verifies `derived_cursor.last_seq` matches `bus.currentSeq` after publish
  - **reconcile replays missed pr.merged → transitions QA work item to done** — simulates crash (pr.merged in log, no derived publisher ran), new publisher reconciles → `phase.changed` emitted, work item moves to `done`
  - **reconcile is a no-op when cursor is caught up (idempotent)** — second publisher reconciles 0 events
  - **reconcile handles multiple missed events** — 2 PRs merged while down, both replayed on restart
  - **reconcile without eventLog returns 0** — safe no-op when eventLog not provided
  - **cursor persists across publisher instances** — second publisher only replays events added after first disposed
- All 21 prior publisher/rule tests continue to pass

### Test Results

CI `check` job passed in 3m50s — typecheck, lint, non-daemon tests, and daemon tests all green.

### CI Status

✓ All required checks passed on run [24877344341](https://github.com/theshadow27/mcp-cli/actions/runs/24877344341):
- `check` ✓ in 3m50s
- `coverage` ✓ in 6m41s
- `build` ✓ in 8s

### PR Comment Surfaces

All 4 surfaces clean: no PR body comments, no inline file:line comments, no review containers, no linked-issue comments.

### Remaining Work (non-blocking)

None identified. The issue's "bonus" cursor generalization is fully implemented.

### Outcome

**qa:pass** — All described functionality implemented and verified. `derived_cursor` table, atomic cursor advancement, startup `reconcile()`, and daemon wiring are all present. 6 new reconciliation tests cover the crash-recovery and downtime scenarios. Tests and CI green.

File: packages/daemon/src/derived-events.ts [hunk 1]
```diff
@@ -3,48 +3,102 @@ import type { MonitorEvent, MonitorEventInput } from "@mcp-cli/core";
...
+   * Replay events from the event log that were missed during downtime or a crash.
+   * Returns the number of events replayed.
+   */
+  reconcile(): number {
+    if (!this.eventLog) return 0;
+    const events = this.eventLog.getSince(this.cursor);
+    for (const event of events) {
+      this.handleEvent(event);
+    }
+    return events.length;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/derived-events.ts]: Fixed in 54ea5c22. `reconcile()` now snapshots `targetSeq = eventLog.currentSeq()` before replay and passes it through `reconcileTarget`. `saveCursor()` caps advancement at `Math.min(seq, reconcileTarget)` so derived events with higher seq cannot leap the cursor past unreplayed original events. Added test "cursor does not leap past unreplayed events when derived events have higher seq".

[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/derived-events.ts]: Fixed in 54ea5c22. `reconcile()` now loops `getSince(this.cursor)` until either the batch is empty or the last event reaches `targetSeq`. This handles >1000 event backlogs without leaving the daemon partially unreconciled.

File: packages/daemon/src/derived-events.ts [hunk 2]
```diff
@@ -3,48 +3,102 @@ import type { MonitorEvent, MonitorEventInput } from "@mcp-cli/core";
...
+  }
+
   private handleEvent(event: MonitorEvent): void {
     const causedBy = event.causedBy ?? [];
     if (causedBy.length >= MAX_DERIVED_DEPTH) return;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/derived-events.ts]: Fixed in 54ea5c22. `handleEvent()` now calls `advanceCursor(event.seq)` before returning on depth-cap, wrapping `saveCursor` in its own transaction. Depth-capped events advance the cursor without applying rules, so reconciliation makes forward progress. Added test "depth-capped events still advance the cursor" which verifies the cursor advances and a second reconcile is a no-op.

File: packages/daemon/src/derived-events.ts [hunk 3]
```diff
@@ -3,48 +3,102 @@ import type { MonitorEvent, MonitorEventInput } from "@mcp-cli/core";
...
+    rules: DerivedRule[];
+    workItemDb: WorkItemDb;
+    db: Database;
+    eventLog?: EventLog;
+  }) {
     this.bus = opts.bus;
     this.rules = opts.rules;
     this.ctx = { workItemDb: opts.workItemDb, bus: opts.bus };
     this.db = opts.db;
+    this.eventLog = opts.eventLog ?? null;
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/derived-events.ts]: Fixed in 54ea5c22. Constructor now reads `bus.eventLog` and: (1) throws if both `opts.eventLog` and `bus.eventLog` are set and differ, (2) defaults `this.eventLog` to `bus.eventLog` when `opts.eventLog` is omitted. This prevents the silent-no-op footgun and means callers who pass `EventBus(eventLog)` get reconciliation for free without remembering to also pass `eventLog`. Added tests for both the mismatch throw and the default-from-bus behavior.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1712 — feat(monitor): startup reconciliation for derived events via cursor replay (fixes #1667)
**Linked issue:** #1667

### Implementation Evidence

- `packages/daemon/src/derived-events.ts`: `DerivedEventPublisher` adds a `derived_cursor` SQLite table (`migrateCursor`/`loadCursor`/`saveCursor`) — cursor is advanced atomically inside the same `db.transaction()` as rule application in `handleEvent`, closing the crash window described in the issue. `reconcile()` snapshots `targetSeq = eventLog.currentSeq()` before replay and loops `getSince(this.cursor)` until caught up, handling >1000 events correctly. Depth-capped events call `advanceCursor(event.seq)` before returning so they don't stall the cursor. Constructor throws on mismatched `eventLog`/`bus.eventLog` and defaults `this.eventLog` from `bus.eventLog` when not explicit.
- `packages/daemon/src/index.ts` (lines 940–950): `eventLog` is now passed into `DerivedEventPublisher`; `reconcile()` is called immediately after construction with an info log on replay.

### Test Coverage

- `packages/daemon/src/derived-events.spec.ts` — 10 new cursor/reconciliation tests covering:
  - Cursor advances as events are processed
  - Reconcile replays missed `pr.merged` → `phase.changed` (crash recovery)
  - Reconcile is a no-op when cursor is caught up (idempotent)
  - Multiple missed events all replayed
  - Reconcile without eventLog returns 0
  - Cursor persists across publisher instances
  - Cursor does not leap past unreplayed events when derived events have higher seq (Copilot fix)
  - Depth-capped events still advance the cursor (Copilot fix)
  - Constructor throws on mismatched eventLog (Copilot fix)
  - Constructor defaults eventLog from bus.eventLog (Copilot fix)
- All 21 prior publisher/rule tests continue to pass

### Test Results

CI run [24877814930](https://github.com/theshadow27/mcp-cli/actions/runs/24877814930): all 3 required jobs passed (check ✓, coverage ✓, build ✓).

### Copilot Inline Comment Surfaces

All 4 Copilot inline threads (cursor leap, >1000-event batch, depth-cap cursor skip, eventLog footgun) have `theshadow27` replies citing fix commit `54ea5c22`. Thread audit: `gh api …/pulls/1712/comments --jq '…' → []` (zero unanswered threads).

### CI Status

✓ All required checks passed on run [24877814930](https://github.com/theshadow27/mcp-cli/actions/runs/24877814930)

### Remaining Work (non-blocking)

None. The issue's bonus cursor generalization is fully implemented.

### Outcome

**qa:pass** — All described functionality implemented and verified. `derived_cursor` table, atomic cursor advancement, startup `reconcile()`, and daemon wiring are all present. 10 new reconciliation/cursor tests cover crash recovery, downtime replay, idempotency, and all four Copilot-flagged edge cases. Tests and CI green.
